### PR TITLE
[venstarthermostat] Initial contribution: Venstar thermostats with Local API support

### PR DESCRIPTION
**Pull request by hww3**
Support for Venstar thermostats with Local API (such as ColorTouch, Voyager and Explorer). This binding includes device autodiscovery and full read/write support using the Venstar local API.

Includes feedback and contributions (as noted in the source) from @digitaldan.

This contribution is ready for review.

Signed-off-by: William Welliver <william@welliver.org>

### CODE REVIEW

**Comment by digitaldan** [comment]
Hi @hww3 , is there anything i can help with in getting the build to pass?

**Comment by hww3** [comment]
Hi @digitaldan- the primary complaint seems to be that it doesn't like my copyright notice.

**Comment by digitaldan** [comment]
Sounds like an easy fix!

**Comment by digitaldan** [comment]
Sounds like an easy fix!

**Comment by digitaldan** [comment]
@hww3 some conflicts have risen up here, can you fix those ?

**Comment by cpmeister** [comment]
This PR contains a lot of changes unrelated to this binding, please try rebasing against the latest 2.5.x branch

**Comment by digitaldan** [comment]
@hww3 any update? Anything i can do ?

**Comment by digitaldan** [comment]
@hww3,  ping :-)

**Comment by digitaldan** [comment]
@hww3 let me know if i can help with anything.

**Comment by hww3** [comment]
@digitaldan -

I seem to have messed this up trying to bring the branch up to date. I haven't had much time to deal with it lately; I also am working with a user on supporting a thermostat model that the current code doesn't work with. That's waiting on a response venstar engineering so I've been trying to get that sorted out first.

**Comment by hww3** [comment]
Hello @digitaldan-

I think I am going to have to drop this PR. It's gotten messed up and trying to fix that only seems to have made it worse.

**Comment by digitaldan** [comment]
Hi @hww3 !

>I think I am going to have to drop this PR. 

After fixing with a rebase, reset, or manually cherry picking the binding code to the 2.5 branch, you can force push to the branch here and not have to close the PR.

> That's waiting on a response venstar engineering so I've been trying to get that sorted out first.

I'm guessing thats not a radical change to the binding, so it would be good to get what we have merged in and do another PR when that change is ready.

>I haven't had much time to deal with it lately

Do you have time to see this PR through review?  If not, I am willing to help out, I would love to see this released in an upcoming 2.5 release, and 3.0 is not that far away, so would love to see it included before that.

**Comment by hww3** [comment]
Hi @digitaldan-

I'll give it another try this week. I'll let you know if there ends up being a problem I can't resolve on my own.

File: bundles/org.openhab.binding.venstarthermostat/pom.xml
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/pom.xml:5]: This can be removed

File: bundles/org.openhab.binding.venstarthermostat/pom.xml
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/pom.xml:29]: OH now ships with jetty 9.4.20, so we can remove this dependency block

File: bundles/org.openhab.binding.venstarthermostat/src/main/feature/feature.xml
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/src/main/feature/feature.xml:3]: This is causing the build failure, and we don't need the jetty line, whole file should be replaced with:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<features name="org.openhab.binding.ventarthermostat-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>

	<feature name="openhab-binding-ventarthermostat" description="Venstar Thermostat Binding" version="${project.version}">
		<feature>openhab-runtime-base</feature>
		<bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.venstarthermostat/${project.version}</bundle>
	</feature>
</features>
```

File: bundles/pom.xml
Comment (digitaldan) [inline bundles/pom.xml:94]: This should not be removed

File: bundles/pom.xml
Comment (digitaldan) [inline bundles/pom.xml:132]: This should not be removed

File: bundles/pom.xml
Comment (digitaldan) [inline bundles/pom.xml:175]: This should not be removed

File: bundles/pom.xml
Comment (digitaldan) [inline bundles/pom.xml:222]: This should not be removed

File: features/pom.xml
Comment (digitaldan) [inline features/pom.xml:1]: This file should not be touched by this PR

File: bundles/org.openhab.binding.venstarthermostat/pom.xml
Comment (digitaldan) [reply bundles/org.openhab.binding.venstarthermostat/pom.xml:5]: to be more specific, the whole build block can be removed

File: bundles/org.openhab.binding.venstarthermostat/pom.xml
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/pom.xml:6]: This can be removed

Comment (hww3) [reply bundles/org.openhab.binding.venstarthermostat/pom.xml:6]: Hi @digitaldan- 

Thanks so much for the tips. I'll incorporate those and push an update. Stay tuned!

**Comment by hww3** [comment]
Hi all-

I think I've resolved all of the outstanding build issues and previous review points. Anyone care to weigh in?

Thanks!

**Comment by digitaldan** [comment]
I will try and take a look this weekend!

File: bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/discovery/VenstarThermostatDiscoveryService.java
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/discovery/VenstarThermostatDiscoveryService.java:242]:
```suggestion
        result = DiscoveryResultBuilder.create(thingUid).withLabel(label).withRepresentationProperty(VenstarThermostatBindingConstants.PROPERTY_UUID)
```

I did notice this one recently as my 3.x system is complaining about it this property if its missing/wrong.

**Review by Hilbrand — CHANGES_REQUESTED** [review]
This binding looks pretty much ready to be merged. It would be great if we can get it in the last 2.5.9 release this weekend (20 September 2020)! Besides some readme comments and the following missing file (changes): This binding should be added to the `CODEOWNERS` file, the `bom/openhab-addons/pom.xml` and a `NOTICE` file must be added. And the one comment from Dan. @digitaldan anything to add?

File: bundles/org.openhab.binding.venstarthermostat/pom.xml
Comment (Hilbrand) [inline bundles/org.openhab.binding.venstarthermostat/pom.xml:9]:
```suggestion
        <version>2.5.9-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:166]: nit - remove empty line

File: bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:173]: nit - remove empty line

File: bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java
Comment (digitaldan) [inline bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:177]: nit - remove empty line

**Review by digitaldan — CHANGES_REQUESTED** [review]
The only issue i see is the discovery suggested change that needs to be fixed and a couple super small nit picks, otherwise LGTM!  FYI i have been running this with 5 thermostats for over a year (2 years?) without a single issue.  Very excited to see it merged.

**Comment by digitaldan** [comment]
@hww3 If we can get this merged today, it can make in it the release tomorrow, otherwise there is a freeze going to happen for a while as we move to 3.0, would be great to get this in so it will be available for both 2.x and 3.x immediately.

**Comment by digitaldan** [comment]
@Hilbrand if  we merge this in now, i can go ahead and do a quick PR to fix the discovery issue.

**Comment by kaikreuzer** [comment]
@digitaldan So your suggestion is to merge and you do immediately a PR wit the outstanding required changes? This would need a lot of trust in the quality of your PR, but considering the special situation, I'd be ok with it.

**Comment by Hilbrand** [comment]
The discovery isn't necessarily a short term breaking change. Besides missing some changes to some files it also has the wrong version in the pom.xml which would break the build. The latter can be fixed by committing my suggestion. If Dan fixes all other comments I'm ok with it.

**Comment by digitaldan** [comment]
As @Hilbrand pointed out, the requested changes are very small, and the code was mergeable otherwise.  I actually never noticed the discovery issue until i ran 3.0, and i have a branch where i have fixed this and have been running for a week, so if we are good with this PR otherwise, then i think its safe with or without a followup PR, but its a simple on liner change that has been fairly well tested by me.

**Comment by kaikreuzer** [comment]
Well, go ahead then, @digitaldan!

**Comment by digitaldan** [comment]
on it !

**Pull request merged**
